lib/os_io.c

```diff
@@ -8901,6 +8901,258 @@ ___SCMOBJ options;)
 
 /*   -   -   -   -   -   -   -   -   -   -   -   -   -   -   -   -   -   -   */
 
+/* Raw device file descriptors */
+typedef struct ___device_raw_struct
+  {
+    ___device base;
+#ifdef USE_POSIX
+    int fd;
+#endif
+  } ___device_raw;
+
+typedef struct ___device_raw_vtbl_struct
+  {
+    ___device_vtbl base;
+  } ___device_raw_vtbl;
+
+
+___HIDDEN int ___device_raw_kind
+   ___P((___device *self),
+        (self)
+___device *self;)
+{
+  return ___RAW_DEVICE_KIND;
+}
+
+___HIDDEN ___SCMOBJ ___device_raw_close_virt
+   ___P((___device *self,
+         int direction),
+        (self,
+         direction)
+        ___device *self;
+        int direction;)
+{
+  ___device_raw *d = ___CAST(___device_raw*,self);
+  int is_not_closed = 0;
+
+  if (d->base.read_stage != ___STAGE_CLOSED)
+    is_not_closed |= ___DIRECTION_RD;
+
+  if (d->base.write_stage != ___STAGE_CLOSED)
+    is_not_closed |= ___DIRECTION_WR;
+
+  if (is_not_closed == 0)
+    return ___FIX(___NO_ERR);
+
+  if ((is_not_closed & ~direction) == 0)
+    {
+      d->base.read_stage = ___STAGE_CLOSED; /* avoid multiple closes */
+      d->base.write_stage = ___STAGE_CLOSED;
+
+#ifdef USE_POSIX
+      if (___close_no_EINTR (d->fd) < 0)
+        return err_code_from_errno ();
+#endif
+    }
+  else if (is_not_closed & direction & ___DIRECTION_RD)
+    d->base.read_stage = ___STAGE_CLOSED;
+  else if (is_not_closed & direction & ___DIRECTION_WR)
+    d->base.write_stage = ___STAGE_CLOSED;
+
+  return ___FIX(___NO_ERR);
+}
+
+___HIDDEN ___SCMOBJ ___device_raw_select_virt
```

Comment:
To clarify: the direction is still in, as they are waitable for input and have an open direction, but they can't be predefined ports because you can't do actual reads.